### PR TITLE
More validation tests

### PR DESCRIPTION
**Pull request by RalfJung**
Still not hard enough, but well, we do what we can.

Also, more tests.

### CODE REVIEW

**Comment by RalfJung** [comment]
I don't understand this last test failure. The same test works just fine on my machine, also using the latest nightly.

**Comment by RalfJung** [comment]
Ah, it's the fullmir run that fails. Got it.

File: src/librustc_mir/interpret/validation.rs [hunk 1]
```diff
@@ -175,7 +175,7 @@ impl<'a, 'tcx, M: Machine<'tcx>> EvalContext<'a, 'tcx, M> {
                 }
             }
 
-            let result = self_.resolve_type_vars_if_possible(result);
+            let result = self_.resolve_type_and_region_vars_if_possible(result);
```
Comment (eddyb) [inline src/librustc_mir/interpret/validation.rs]: This doesn't do anything without that other line I gave you, which you have to place just above this one.

Comment (RalfJung) [inline src/librustc_mir/interpret/validation.rs]: You mean the one I removed in <https://github.com/solson/miri/pull/301/commits/c39bc68f6b0c714f7a9cccf76f405be3a0a5a1a2> because it broke some testcases?

Comment (eddyb) [inline src/librustc_mir/interpret/validation.rs]: Ah I didn't see that :/. Welp, that's what happens when there are leftover `ReErased` in the *input* to this.

Comment (RalfJung) [inline src/librustc_mir/interpret/validation.rs]: ... wherever these are coming from.

Comment (RalfJung) [inline src/librustc_mir/interpret/validation.rs]: So should I be revert the change to `type_and_region_vars` as well? (And then I guess squash the commits to avoid needless reverts; GitHub insists on a merge/rebase anyway.)

**Comment by RalfJung** [comment]
@eddyb can we merge this or do you want me to change `type_and_region_vars` back to `type_vars`?

**Comment by eddyb** [comment]
Do whatever you want but know that this PR does *nothing* other than add two tests.

**Comment by RalfJung** [comment]
Build is currently broken by <https://github.com/rust-lang/rust/issues/43973>. I am working on a PR to fix that.

**Comment by RalfJung** [comment]
#305 I keep getting spurious errors like
```
error: compilation failed!

status: exit code: 101

command: target/release/miri tests/run-pass/bad_substs.rs -L /tmp --target=x86_64-unknown-linux-gnu -Z dump-mir=all -Z mir-opt-level=3 -Z dump-mir-dir=/tmp/bad_substs -L /tmp/bad_substs.stage-id.mir-opt.libaux -C prefer-dynamic -o /tmp/bad_substs.stage-id -Zmir-opt-level=0 -Zmir-emit-validate=1

stdout:

------------------------------------------

------------------------------------------

stderr:

------------------------------------------

error: linking with `cc` failed: exit code: 1

  |

  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/bad_substs.0.o" "-o" "/tmp/bad_substs.stage-id" "/tmp/bad_substs.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-L" "/tmp" "-L" "/tmp/bad_substs.stage-id.mir-opt.libaux" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-l" "std-bd7febf8d59547f4" "-Wl,-Bstatic" "/home/travis/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-b8ac48e67742a8a2.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"

  = note: cc: error: /tmp/bad_substs.0.o: No such file or directory

          cc: error: /tmp/bad_substs.crate.allocator.o: No such file or directory
```
